### PR TITLE
Custom Type Conversion - load_graphml

### PR DESCRIPTION
**Pull request by d-wasserman**
This PR addresses issue #575, and willl enable edges and nodes imported with load_graphml to be able to do type conversions for attributes beyond the default attributes currently being converted to floats. This enables users who build graphs with additional edge weights or node attributes that can be properly saved and then loaded based on the desired attribute data type. 

Relevant Code Snippets

```
def load_graphml(filepath, node_type=int, node_dtypes=None, edge_dtypes=None):
    """
    Load an OSMnx-saved GraphML file from disk.

    Converts the node/edge attributes to appropriate data types.

    Parameters
    ----------
    filepath : string
        path to the GraphML file
    node_type : type
        convert node ids to this data type
    node_dtypes : dict of attribute name -> data type
        identifies additional is a numpy.dtype or Python type to cast one or more additional node attributes
        defaults to {"elevation":float, "elevation_res":float, "lat":float, "lon":float, "x":float, "y":float} if None
    edge_dtypes : dict of attribute name -> data type
        identifies additional is a numpy.dtype or Python type to cast one or more additional edge attributes. Defaults
        to {"length": float, "grade": float, "grade_abs": float, "bearing": float, "speed_kph": float,
       "travel_time": float} if None
    Returns
    -------
    G : networkx.MultiDiGraph
    """
    # read the graph from disk
    G = nx.MultiDiGraph(nx.read_graphml(filepath, node_type=node_type))

    # convert node/edge attribute data types
    utils.log("Converting node and edge attribute data types")
    G = _convert_node_attr_types(G, node_type, node_dtypes)
    G = _convert_edge_attr_types(G, node_type, edge_dtypes)
```

### CODE REVIEW

**Comment by gboeing** [comment]
Thanks. Looks like the tests are failing... Check the Travis CI logs for details.

**Comment by d-wasserman** [comment]
It seems to be a doc string test with a bad indentation. I assume you are using something like McCabe index for code complexity and that got triggered too. It seems the other tests passed.

**Comment by d-wasserman** [comment]
After these doc string edits, I am not sure really what to do on the complexity of the edge converter function. It was already pretty complex given the edge cases it was dealing with, and I am not sure what put it over the edge beyond the new parameter. I will leave it to your discretion to determine a potential path to address this. 

I will address any doc string tests should they come up after these changes.

**Comment by d-wasserman** [comment]
The only test I see failing is this one now:
"./osmnx/io.py:260:1: C901 '_convert_edge_attr_types' is too complex (16)"
The function was already pretty complex and I did attempt to simplify it by incorporating length into the general converter. Let me know if you want any further edits.

File: osmnx/io.py [hunk 1]
```diff
@@ -353,15 +365,15 @@ def _stringify_nonnumeric_cols(gdf):
...
-    edge_tag_aggs=None,
+        data,
+        filepath=None,
+        node_tags=settings.osm_xml_node_tags,
+        node_attrs=settings.osm_xml_node_attrs,
+        edge_tags=settings.osm_xml_way_tags,
+        edge_attrs=settings.osm_xml_way_attrs,
+        oneway=False,
+        merge_edges=True,
+        edge_tag_aggs=None,
```
[the remark this answers is not in the document]
Comment (d-wasserman) [reply osmnx/io.py]: Did not mean to touch that function. Should have paid more attention after reformatting. Thanks for catching that.

**Comment by d-wasserman** [comment]
@Labulitiolle it seems there is still some complexity test still being an issue.

File: osmnx/io.py [hunk 2]
```diff
@@ -577,7 +600,7 @@ def _append_edges_xml_tree(root, gdf_edges, edge_attrs, edge_tags, edge_tag_aggs
             else:
                 for tag in edge_tags:
                     if (tag in all_way_edges.columns) and (
-                        tag not in (t for t, agg in edge_tag_aggs)
+                            tag not in (t for t, agg in edge_tag_aggs)
```
Comment (Labulitiolle) [inline osmnx/io.py:603]: Still an unnecessary white space here. Hope this will make it

File: osmnx/io.py [hunk 3]
```diff
@@ -250,17 +267,23 @@ def _convert_edge_attr_types(G, node_type):
...
     Returns
     -------
     G : networkx.MultiDiGraph
     """
-    # convert numeric, bool, and list edge attributes from string to correct data types
+    if edge_dtypes is None:
+        edge_dtypes = {"length": float, "grade": float, "grade_abs": float,
+                       "bearing": float, "speed_kph": float, "travel_time": float}
+    # convert numeric, bool, and list edge attributes from string
+    # to correct data types
```
Comment (Labulitiolle) [inline osmnx/io.py:285]: By adding this `if` condition, you increased the complexity from 15 to 16 ([link](https://www.flake8rules.com/rules/C901.html)). As I don't see an obvious way to reduce the complexity of `_convert_edge_attr_types` I see two ways to bypass this test:

1. Increase the `flake8` complexity threshold (Not sure how to do that)
2. Split the `_convert_edge_attr_types` function.

Comment (d-wasserman) [reply osmnx/io.py:285]: I find it funny an urban planning inspired library has the same "last one in pays" issues faced by real estate developers in redevelopment markets. Software developers were naïve to think they were immune!
I vote to increase the flake8 complexity slightly here. The function was already complex, and I think I already contributed to make it more readable. I am not sure breaking it up further makes sense to a human reading it. Including the doc string I might have to add another 6-10 lines to the code base to accommodate the test. I am also content to just sit on the PR for a while to see if any consensus develops.

Comment (gboeing) [reply osmnx/io.py:285]: I'll have some time to take a look in the coming couple of days. I can make some revisions as needed from there. Thanks for the contribution.

Comment (d-wasserman) [reply osmnx/io.py:285]: Sounds good! Thanks Geoff!
